In [23]:
# basic functions
import math
import numpy as np

In [24]:
#visual style
import matplotlib.pyplot as plt

%matplotlib inline

## `Exercises`

*1. A friend finds a coin on the ground, flips it, and gets six heads in a row and then one tails. Give the beta distribution that describes this. Use integration to determine the probability that the true rate of flipping heads is between 0.4 and 0.6, reflecting that the coin is reasonably fair.*

The **Beta Distibution** parameters are **`alpha (observed successes)`** and **`beta (observed failures)`**. Problems using this distribution calculate the rate of success given the data we have.

P(RateOfSuccess|Successes and Failures) = Beta($\alpha$,$\beta$)

For this problem the **Number of Heads = 6** (ie "successes") and the number of failures equals the **Number of Tails = 1**

With this data we can determine the chances that the coin is fair or "that the true rate of flipping heads is between 0.4 and 0.6."

To solve this using Python, use **`from scipy.stats import beta`**.

Another way of thinking about this is "What is the probability the rate of heads is between 0.4 and 0.6?" 

To do that, look at the chunk of probability that falls between those two bookends using the cumulative distribution method or **`cdf`** method.

In [128]:
from scipy.stats import beta

In [130]:
# parameters
heads_q1 = 6
tails_q1 = 1
fair_rate = [0.4,0.6]

In [131]:
result = beta.cdf(fair_rate,heads_q1,tails_q1)
result

array([0.004096, 0.046656])

The result is the cumulative distribution up to the lower bound (0.4) and up to the upper bound (0.6). To get the probability that the true rate of heads falls below that, subtract the lower bound output from the upper bound.

*Note: If either alpha or beta is 0, the result will be nan.*

In [48]:
result[-1] - result[0]

0.04255999999999999

In [270]:
def beta_prob(a=1,b=1,bounds=[0.0,1.0]):
    """
    Calculates beta probability between a lower and upper bound.
    
    Parameters
    ----------
    a: int 
        alpha
        
    b: int
        beta
        
    bounds: list
        lower bound,upper bound
    
    Returns
    -------
    out: np.float64
        probability
    """
    result = beta.cdf(bounds,a,b)
    return result[-1] - result[0]

In [271]:
beta_prob(6,1,[0.4,0.6])

0.04255999999999999

*2. Come up with a prior probability that the coin is fair. Use a beta distribution such that there is at least a 95 percent chance that the true rate of flipping heads is between 0.4 and 0.6.*

Another way to phrase this question would be:

What prior probability do you need to determine we are 95% the true rate of flipping heads is between 0.4 and 0.6?

As Will Kurt points out in the answer, "Any $\alpha$<sub>prior</sub> = $\beta$<sub>prior</sub> will give a 'fair' prior; and the larger those values are, the stronger that prior is."

In other words, for a fair prior the successes should equal the failures. And if you take successes plus failures, the higher the number, the better.

In [90]:
beta_prob(fair_rate,26,21)

0.7208847847634349

In [97]:
steps = [n for n in np.arange(0,101,10)]
steps

[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

In [106]:
for s in steps:
    n = s/2
    print(f'{s}: {beta_prob(fair_rate,6+n,1+n):.2f}')

0: 0.04
10: 0.31
20: 0.50
30: 0.63
40: 0.72
50: 0.79
60: 0.84
70: 0.87
80: 0.90
90: 0.92
100: 0.94


In [116]:
steps_2 = [n for n in np.arange(107,111,1)]
steps_2

[107, 108, 109, 110]

In [117]:
for s in steps_2:
    n = s/2
    print(beta_prob(fair_rate,6+n,1+n))

0.9492233998960515
0.9504274807668823
0.9516017214580128
0.9527469094270735


In [189]:
def beta_prior_finder(bounds=[0.0,1.0], a=1,b=1,start=0,finish=10,step=1):
    """
    work in progress, starting with 'fair' prior
    really, fair_beta_prior_finder
    if 'fair', prior returned is divided equally amongst alpha and beta
    returns confidence level for certain prior n
    returns (prior, prob)
    """
    prior_range = [n for n in np.arange(start,finish,step)]
    return [(n,beta_prob(bounds,a+n/2,b+n/2)) for n in prior_range]

In [160]:
beta_prior_finder(bounds=[0.4,0.6],a=6,b=1,start=107,finish=111,step=1)

[(107, 0.9492233998960515),
 (108, 0.9504274807668823),
 (109, 0.9516017214580128),
 (110, 0.9527469094270735)]

So this says we'd need a prior of 108 to be 95% certain

In [269]:
x = np.arange(1,7,1)

fig,ax = plt.subplots()
ax.plot(x,beta_pdf
plt.show()

SyntaxError: invalid syntax (<ipython-input-269-088533af84dc>, line 5)

In [239]:
y.shape

(2,)

## `Notes`

Think of some binary outcomes where this might be useful in your life.